app/status.go

```diff
@@ -281,6 +282,36 @@ func (a *App) SetStatusAwayIfNeeded(userId string, manual bool) {
 	a.SaveAndBroadcastStatus(status)
 }
 
+// SetStatusDoNotDisturbTimed takes endtime in RFC3339 string format in UTC
+// and sets status of given userId to dnd which will be restored back after endtime
+func (a *App) SetStatusDoNotDisturbTimed(userId string, endtime string) {
+	if !*a.Config().ServiceSettings.EnableUserStatuses {
+		return
+	}
+
+	status, err := a.GetStatus(userId)
+
+	if err != nil {
+		status = &model.Status{UserId: userId, Status: model.STATUS_OFFLINE, Manual: false, LastActivityAt: 0, ActiveChannel: ""}
+	}
+
+	status.PrevStatus = status.Status
+	status.Status = model.STATUS_DND
+	status.Manual = true
+
+	// take a string in RFC3339 format and return time
+	utc, cErr := time.Parse(time.RFC3339, endtime)
+
+	if cErr != nil {
+		mlog.Error("Failed to parse endtime", mlog.String("user_id", userId), mlog.String("err", cErr.Error()))
+		return
+	}
```

Comment:
This validation sould happen even before of doing the `GetStatus` that way, any wrong request doesn't need to execute the DB query.